### PR TITLE
Add support for cloudrun ports

### PR DESCRIPTION
**Pull request by dj80hd**
Fixes: terraform-providers/terraform-provider-google#5539

API Reference:
https://cloud.google.com/run/docs/reference/rest/v1/RevisionSpec#ContainerPort
https://cloud.google.com/run/docs/reference/rest/v1/RevisionSpec#Container

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
cloudrun: added `ports` field to `google_cloud_run_service` `templates.spec.containers`
```

### CODE REVIEW

**Comment by dj80hd** [comment]
After more thinking and testing, I realized this solution may not be backwards compatible.

**Comment by slevenick** [comment]
> After more thinking and testing, I realized this solution may not be backwards compatible.

Why is that? Adding a new field shouldn't be a problem

**Comment by dj80hd** [comment]
> Why is that? Adding a new field shouldn't be a problem

Sorry, false alarm.  All tests pass as expected now.

File: products/cloudrun/api.yaml [hunk 1]
```diff
@@ -509,6 +509,21 @@ objects:
...
                     exists or not.
                     Defaults to "".
+            - !ruby/object:Api::Type::Array
+              name: ports
+              description: |-
+                List of open ports in the container.
+              item_type: !ruby/object:Api::Type::NestedObject
+                properties:
+                - !ruby/object:Api::Type::String
+                  name: name
```
[the remark this answers is not in the document]
Comment (dj80hd) [reply products/cloudrun/api.yaml:519]: I wasn't sure `name` and `protocol` should be here at all.   
The original suggestion in the issue is to only allow `container_port`.   
I figured these are just kubernetes values under the hood.

[the remark this answers is not in the document]
Comment (dj80hd) [reply products/cloudrun/api.yaml:519]: On cloudrun fully managed, setting `name` and `protocol` results in 400 errors like this.
```
Error: Error updating Service "locations/us-west1/namespaces/some-project/services/some-name-cloud-run": googleapi: Error 400: The request has errors
Details:
[
  {
    "@type": "type.googleapis.com/google.rpc.BadRequest",
    "fieldViolations": [
      {
        "description": "spec.template.spec.containers[0].ports[0].name should be empty",
        "field": "spec.template.spec.containers[0].ports[0].name"
      }
    ]
  }
]
```
This behavior is consistent with the api documentation: https://cloud.google.com/run/docs/reference/rest/v1/RevisionSpec#ContainerPort 

If this resource supports both cloudrun fully managed and cloudrun for anthos, i think optional is better. 
```

**Comment by kurkop** [comment]
Great!, I'm waiting for these fields.

File: products/cloudrun/api.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (slevenick) [reply products/cloudrun/api.yaml:521]: It is probably worth adding `required` to `containerPort` to enforce that a user can't specify an empty block here. It seems like `containerPort` will always need to be specified to be useful?

[the remark this answers is not in the document]
Comment (dj80hd) [reply products/cloudrun/api.yaml:521]: containerPort is now required.

**Comment by dj80hd** [comment]
> It looks like this change is causing tests to fail

Can i run these tests locally?  I made the suggested change but am not sure if i am testing correctly.

My current test method is:

1. Generate provider code with `bundle exec compiler -p "products/cloudrun" -v "ga" -e terraform -o "$GOPATH/src/github.com/terraform-providers/terraform-provider-google"`

2. Build the provider with `make build test`

3. Run acceptance test in provider directory: `make testacc TEST=./google TESTARGS='-run TestAccCloudRunService_cloudRunServiceUpdate'`

4. Do some real terraform sanity tests using locally built provider

I did notice some lint errors unrelated to my change.

I also had to manually revert some unrelated, unexpected import changes [like these](https://github.com/terraform-providers/terraform-provider-google/pull/6808).   
The diff report from modular-magician looked fine, however, so i just ignored this.

**Comment by slevenick** [comment]
> > It looks like this change is causing tests to fail
> 
> Can i run these tests locally? I made the suggested change but am not sure if i am testing correctly.
> 
> My current test method is:
> 
> 1. Generate provider code with `bundle exec compiler -p "products/cloudrun" -v "ga" -e terraform -o "$GOPATH/src/github.com/terraform-providers/terraform-provider-google"`
> 2. Build the provider with `make build test`
> 3. Run acceptance test in provider directory: `make testacc TEST=./google TESTARGS='-run TestAccCloudRunService_cloudRunServiceUpdate'`
> 4. Do some real terraform sanity tests using locally built provider
> 
> I did notice some lint errors unrelated to my change.
> 
> I also had to manually revert some unrelated, unexpected import changes [like these](https://github.com/terraform-providers/terraform-provider-google/pull/6808).
> The diff report from modular-magician looked fine, however, so i just ignored this.

Yes, generally that is the way to run the generated tests. I'd suggest running a more broad suite of tests occasionally to make sure that your changes don't impact other uses of the cloud run resource.

`make testacc TEST=./google TESTARGS='-run TestAccCloudRunService'` is what I'm running to check this

**Comment by dj80hd** [comment]
Thank you for your guidance on this first contribution. 
`make testacc TEST=./google TESTARGS='-run TestAccCloudRunService'` now passes all tests.

File: products/cloudrun/api.yaml [hunk 2]
```diff
@@ -509,6 +509,24 @@ objects:
...
+                  description: Name of the port.
+                - !ruby/object:Api::Type::String
+                  name: protocol
+                  description: Protocol used on port. Defaults to TCP.
+                - !ruby/object:Api::Type::Integer
+                  name: containerPort
+                  description: Port number.  Defaults to 8080.
```
Comment (slevenick) [inline products/cloudrun/api.yaml:528]:
```suggestion
                  description: Port number.
```

If this is required it doesn't default to anything

**Pull request merged**